New Node.js static server example (#13113)
mdn/content: files/en-us/learn/server-side/node_server_without_framework/index.md

---
{{LearnSidebar}}

> This article provides a simple static file server built with pure [Node.js](https://nodejs.org/en/) without the use of a framework. Current state of Node.js is such that almost everything we need can be easily solved by powerful internal APIs and a few lines of code.

Reviewer: For readabality "The current state of" is better than "Current state of ...".

```suggestion
This article provides a simple static file server built with pure [Node.js](https://nodejs.org/en/) without the use of a framework. 
The current state of Node.js is such that almost everything we needed is provided by hte inbuilt APIs and just a few lines of code.
```

  png: 'image/png',
  jpg: 'image/jpg',
  gif: 'image/gif',
  ico: 'image/x-icon',
  svg: 'image/svg+xml',
};

const STATIC_PATH = path.join(process.cwd(), './static');

const toBool = [() => true, () => false];

const prepareFile = async (url) => {
  const paths = [STATIC_PATH, url];
  if (url.endsWith('/')) paths.push('index.html');
  const filePath = path.join(...paths);
  const pathTraversal = !filePath.startsWith(STATIC_PATH);
  const exists = await fs.promises.access(filePath).then(...toBool);
  const found = !pathTraversal && exists;
  const streamPath = found ? filePath : STATIC_PATH + '/404.html';
  const ext = path.extname(streamPath).substring(1).toLowerCase();
  const stream = fs.createReadStream(streamPath);
  return { found, ext, stream };
};

http.createServer(async (req, res) => {
  const file = await prepareFile(req.url);
  const statusCode = file.found ? 200 : 404;
  const mimeType = MIME_TYPES[file.ext] || MIME_TYPES.default;
  res.writeHead(statusCode, { 'Content-Type': mimeType });
  file.stream.pipe(res);
  console.log(`${req.method} ${req.url} ${statusCode}`);
}).listen(PORT);

console.log(`Server running at http://127.0.0.1:${PORT}/`);
```

### Breakdown

> Following lines loads internal modules Node.js modules, roughly similar to imports.

Reviewer: Perhaps remove ", roughly similar to imports" - since if they don't know require they probably don't know imports either. If you strongly disagreed, suggest you link to the imports reference page.
```suggestion
The following lines load internal Node.js modules.
```

```js
const fs = require('fs');
const http = require('http');
const path = require('path');
```

> Afterward is the function for creating the server. `https.createServer` returns a `Server` object, which we can start up by listening on `PORT`.

Reviewer: ```suggestion
Next we have a function for creating the server. `https.createServer` returns a `Server` object, which we can start up by listening on `PORT`.
```

```js
http.createServer((req, res) => {
  /* handle http requests */
}).listen(PORT);

console.log(`Server running at http://127.0.0.1:${PORT}/`);
```

> Async function `prepareFile` returns structure: `{ found: boolean , ext: string, stream: ReadableStream }` If file can be served (server process have an access and no path-traversal vulnerability found), we will return `200` as a `statusCode`, othervise `404`. Note that otehr status codes can be found in `http.STATUS_CODES`. With `404` status we will return content of `'/404.html'` file.

Reviewer: ```suggestion
The asynchronous function `prepareFile` returns the structure: `{ found: boolean , ext: string, stream: ReadableStream }`.
If the file can be served (the server process has access and no path-traversal vulnerability is found), we will return the HTTP status of `200` as a `statusCode` indicating success (otherwise we return `HTTP 404`). 
Note that other status codes can be found in `http.STATUS_CODES`.
With `404` status we will return content of `'/404.html'` file.
```

> Еhe extension of the file being requested will be parsed and lower-cased. After that we will search `MIME_TYPES` collection for the right [MIME types](/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types). If no matches are found, we use the `application/octet-stream` as the default type.

Reviewer: ```suggestion
The extension of the file being requested will be parsed and lower-cased. After that we will search `MIME_TYPES` collection for the right [MIME types](/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types). If no matches are found, we use the `application/octet-stream` as the default type.
```

> Finally, if there are no errors, we send over the requested file. In any case `file.stream` will contain `Readable` stream which wille be piped into `res` (instance of the `Writable` stream).

Reviewer: ```suggestion
Finally, if there are no errors, we send the requested file. The `file.stream` will contain a `Readable` stream that will be piped into `res` (an instance of the `Writable` stream).
```

---
{{LearnSidebar}}

This article provides a simple static file server built with pure [Node.js](https://nodejs.org/en/) without the use of a framework. 
The current state of Node.js is such that almost everything we needed is provided by the inbuilt APIs and just a few lines of code.

## Example

A simple static file server built with Node.js:

<details>

> <summary>CJS implementation</summary>

Reviewer 2: Before we reached resolution with https://github.com/orgs/mdn/discussions/191, could you keep this CJS-only? This is simply confusing without another explanatory document.

Author: @Josh-Cena done

---
{{LearnSidebar}}

This article provides a simple static file server built with pure [Node.js](https://nodejs.org/en/) without the use of a framework. 
The current state of Node.js is such that almost everything we needed is provided by the inbuilt APIs and just a few lines of code.

## Example

A simple static file server built with Node.js:

```js
'use strict';

> const fs = require('fs');

Reviewer 2: Note: as a mid-term goal I'd like MDN Node tutorials to embrace Node ESM. More and more dependencies are migrating to ESM and you can't `require` ESM, which means for forward compatibility all user projects should start as ESM.

Reviewer 3: @Josh-Cena: I think this is a great point. To give it more visibility, can you start a discussion about it (I believe the consensus will be easy to reach) at mdn/mdn-community ?

Reviewer 2: Ah yes, I should—just haven't got to it yet...

Reviewer 2: https://github.com/orgs/mdn/discussions/191
